client/src/hooks/graphql/use-with-import-export.tsx

```diff
@@ -0,0 +1,157 @@
+  const { mentorState } = useActiveMentor();
+  const mentor = mentorState.data;
+
+  function exportMentor() {
+    if (!mentor) {
+      return;
+    }
+    api.exportMentor(mentor._id).then((m) => {
+      const element = document.createElement("a");
+      element.setAttribute(
+        "href",
+        "data:text/json;charset=utf-8," + encodeURIComponent(JSON.stringify(m))
+      );
+      element.setAttribute("download", "mentor.json");
+      element.style.display = "none";
+      document.body.appendChild(element);
+      element.click();
+      document.body.removeChild(element);
+    });
+  }
+
+  function importMentor(file: File) {
+    if (!mentor) {
+      return;
+    }
+    const reader = new FileReader();
+    reader.onload = function (e) {
+      if (typeof e.target?.result === "string") {
+        const json = JSON.parse(e.target?.result);
+        updateImport(json);
+      }
+    };
+    reader.readAsText(file);
+  }
+
+  function updateImport(json: MentorExportJson) {
+    if (!mentor) {
+      return;
+    }
+    setImportJson(json);
+    api.importMentorPreview(mentor._id, json).then((p) => {
+      console.log(JSON.stringify(p, null, " "));
+      setImportPreview(p);
+    });
+  }
+
+  function confirmImport() {
+    if (!importedJson || !mentor) {
+      return;
+    }
+    api.importMentor(mentor._id, importedJson, accessToken).then(() => {
+      setImportJson(undefined);
+      setImportPreview(undefined);
+    });
+  }
+
+  function cancelImport() {
+    setImportJson(undefined);
+    setImportPreview(undefined);
+  }
+
+  function mapSubject(subject: Subject, replacement: Subject) {
+    if (!importedJson || !importPreview || !mentor) {
+      return;
+    }
+    const json = { ...importedJson };
+    // if the replacement subject is already referenced elsewhere in the import, remove it so we don't include it twice
+    const rIdx = json.subjects.findIndex((s) => s._id === replacement._id);
+    ~rIdx && json.subjects.splice(rIdx, 1);
+    // find and replace the "new" subject in import with the replacement subject
+    // is there a case where we want to some changes from import instead of over-writing with replacement?
+    const idx = json.subjects.findIndex((s) => s._id === subject._id);
+    ~idx && json.subjects.splice(idx, 1, replacement);
+    // TODO: if the subject that was replaced had some questions that were overwritten by the replacement
+    // we might need to remap them in imported questions and answers list
+    updateImport(json);
+  }
+
+  function mapQuestion(question: Question, replacement: Question) {
+    if (!importedJson || !importPreview || !mentor) {
+      return;
+    }
+    const json = { ...importedJson };
+    // if the replacement question is already referenced elsewhere in the import, remove it so we don't include it twice
+    let rIdx = json.questions.findIndex((q) => q._id === replacement._id);
+    ~rIdx && json.questions.splice(rIdx, 1);
```

Comment:
all these splices of `json.question` are mutations of state, no?

We need to never mutate the existing state. If you want to simplify and void wasteful/repetitive copying, then just copy one time above, e.g.

```typescript
const json = { 
  ...importedJson,
  questions: ...json?.questions || []
};
```